peer/ping_manager.go

```diff
@@ -0,0 +1,229 @@
+
+	SendPing      func(ping *lnwire.Ping)
+	OnPongFailure func()
+}
+
+// pingManager is a structure that is designed to manage the internal state
+// of the ping pong lifecycle with the remote peer. We assume there is only one
+// ping outstanding at once.
+//
+// NOTE: This structure MUST be initialized with NewPingManager.
+type PingManager struct {
+	cfg *PingManagerConfig
+
+	// pingTime is a rough estimate of the RTT (round-trip-time) between us
+	// and the connected peer.
+	// To be used atomically.
+	// TODO(roasbeef): also use a WMA or EMA?
+	pingTime atomic.Pointer[time.Duration]
+
+	// pingLastSend is the time when we sent our last ping message.
+	// To be used atomically.
+	pingLastSend *time.Time
+
+	// outstandingPongSize is the current size of the requested pong
+	// payload.  This value can only validly range from [0,65531]. Any
+	// value < 0 is interpreted as if there is no outstanding ping message.
+	outstandingPongSize int32
+
+	// pingTicker is a pointer to a Ticker that fires on every ping
+	// interval.
+	pingTicker *time.Ticker
+
+	// pingTimeout is a Timer that will fire when we want to time out a
+	// ping
+	pingTimeout *time.Timer
+
+	// pongChan is the channel on which the pingManager will write Pong
+	// messages it is evaluating
+	pongChan chan *lnwire.Pong
+
+	started sync.Once
+	stopped sync.Once
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewPingManager constructs a pingManager in a valid state. It must be started
+// before it does anything useful, though.
+func NewPingManager(cfg *PingManagerConfig) *PingManager {
+	m := PingManager{
+		cfg:                 cfg,
+		outstandingPongSize: -1,
+		pongChan:            make(chan *lnwire.Pong, 1),
+		quit:                make(chan struct{}),
+	}
+
+	return &m
+}
+
+// Start launches the primary goroutine that is owned by the pingManager.
+func (m *PingManager) Start() error {
+	var err error
+	m.started.Do(func() {
+		err = m.start()
+	})
+
+	return err
+}
+
+func (m *PingManager) start() error {
+	m.pingTicker = time.NewTicker(m.cfg.IntervalDuration)
+
+	m.pingTimeout = time.NewTimer(0)
+	defer m.pingTimeout.Stop()
+
+	if !m.pingTimeout.Stop() {
+		<-m.pingTimeout.C
+	}
+
+	m.wg.Add(1)
+	go func() {
+		defer m.wg.Done()
+		for {
+			select {
+			case <-m.pingTicker.C:
+				if m.outstandingPongSize >= 0 {
+					m.cfg.OnPongFailure()
+					return
+				}
+
+				pongSize := m.cfg.NewPongSize()
+				ping := &lnwire.Ping{
+					NumPongBytes: pongSize,
+					PaddingBytes: m.cfg.NewPingPayload(),
+				}
+
+				if err := m.setPingState(pongSize); err != nil {
+					m.cfg.OnPongFailure()
+
+					return
+				}
+
+				m.cfg.SendPing(ping)
+
+			case <-m.pingTimeout.C:
+				m.resetPingState()
+
+				m.cfg.OnPongFailure()
+
+				return
+
+			case pong := <-m.pongChan:
+				pongSize := int32(len(pong.PongBytes))
+				expected := m.outstandingPongSize
+				lastPing := m.pingLastSend
+				m.resetPingState()
+				if pongSize != expected {
+					m.cfg.OnPongFailure()
```

Comment:
Since `OnPongFailure` is called for a variety of reasons, perhaps it makes sense to give the callback an `error` param so we can provide more info which Brontide can then log.